front/src/applications/operationalStudies/hooks/useLazyLoadTrains.ts

```diff
@@ -0,0 +1,128 @@
+/* eslint-disable no-restricted-syntax, no-await-in-loop */
+import { useEffect, useState, type Dispatch, type SetStateAction, useMemo } from 'react';
+
+import { keyBy, sortBy } from 'lodash';
+import { useSelector } from 'react-redux';
+
+import {
+  osrdEditoastApi,
+  type PathfindingResultSuccess,
+  type TrainScheduleResult,
+} from 'common/api/osrdEditoastApi';
+import { useOsrdConfSelectors } from 'common/osrdContext';
+import useLazyProjectTrains from 'modules/simulationResult/components/SpaceTimeChart/useLazyProjectTrains';
+import { getTrainIdsPackage } from 'modules/simulationResult/components/SpaceTimeChart/utils';
+import type { TrainScheduleWithDetails } from 'modules/trainschedule/components/TimetableV2/types';
+import type { Dictionary } from 'utils/types';
+
+import formatTrainScheduleSummaries from '../helpers/formatTrainScheduleSummaries';
+
+const BATCH_SIZE = 10;
+
+type useLazyLoadTrainsProp = {
+  infraId?: number;
+  trainIdsToFetch?: number[];
+  path?: PathfindingResultSuccess;
+  trainSchedules?: TrainScheduleResult[];
+  setTrainIdsToFetch: Dispatch<SetStateAction<number[] | undefined>>;
+};
+
+/**
+ * This hook gradually fetches and projects trains of the timetable.
+ *
+ * It first fetches the simulation of 10 trains at a time, then projects them on the path.
+ * This optimizes the performance of the application and allow us to display the trains as
+ * soon as they are ready.
+ */
+const useLazyLoadTrains = ({
+  infraId,
+  trainIdsToFetch,
+  path,
+  trainSchedules,
+  setTrainIdsToFetch,
+}: useLazyLoadTrainsProp) => {
+  const { getElectricalProfileSetId } = useOsrdConfSelectors();
+  const electricalProfileSetId = useSelector(getElectricalProfileSetId);
+
+  const [trainScheduleSummariesById, setTrainScheduleSummariesById] = useState<
+    Dictionary<TrainScheduleWithDetails>
+  >({});
+  const [trainIdsToProject, setTrainIdsToProject] = useState<number[]>([]);
+  const [allTrainsLoaded, setAllTrainsLoaded] = useState(false);
+
+  const trainScheduleSummaries = useMemo(
+    () => sortBy(Object.values(trainScheduleSummariesById), 'startTime'),
+    [trainScheduleSummariesById]
+  );
+
+  const [postV2TrainScheduleSimulationSummary] =
+    osrdEditoastApi.endpoints.postV2TrainScheduleSimulationSummary.useLazyQuery();
+
+  const { data: { results: rollingStocks } = { results: [] } } =
+    osrdEditoastApi.endpoints.getLightRollingStock.useQuery({ pageSize: 1000 });
+
+  const trainSchedulesById = useMemo(() => keyBy(trainSchedules, 'id'), [trainSchedules]);
+
+  const { projectedTrains, setProjectedTrainsById } = useLazyProjectTrains({
+    infraId,
+    trainIdsToProject,
+    path,
+    trainSchedules,
+    moreTrainsToCome: !allTrainsLoaded,
+    setTrainIdsToProject,
+  });
+
+  // gradually fetch the simulation of the trains
+  useEffect(() => {
+    const getTrainScheduleSummaries = async (_infraId: number, _trainToFetchIds: number[]) => {
```

Comment:
Hm, that `_infraId` underscore trick is a bit unfortunate. Maybe we should just allow variable shadowing…